File: js/src/either.js
import Monad from './monad';

export class Either extends Monad {
  // pure :: a -> M a
  pure = (value) => {
    return new Right(value)
  }

  // flatMap :: # M a -> (a -> M b) -> M b
  flatMap = f => 
    this.constructor.name === 'Left' ? 
    this : 
    f(this.value)

}

export class Left extends Either {
  constructor(value) {
    super();
    this.value = value;
  }

  toString() {
    return `Left(${this.value})`
  }
}

export class Right extends Either {
  constructor(value) {
    super();
    this.value = value;
  }

  toString() {
    return `Right(${this.value})`
  }
}

// attempt :: (() -> a) -> M a 
Either.attempt = f => {
    try {
      return new Right(f())
    } catch(e) {
      return new Left(e)
    }
}
Either.pure = (new Left(null)).pure

File: swift/Monad/Monad/Either.swift
//
//  Either.swift
//  Monad
//
//  Created by Алексей Карасев on 15/08/2017.
//  Copyright © 2017 Alexey Karasev. All rights reserved.
//

import Foundation

enum Either<A, B> {
  case Left(A)
  case Right(B)
  static func pure<C>(_ value: C) -> Either<A, C> {
    return Either<A, C>.Right(value)
  }
  
  func flatMap<D>(_ f: (B) -> Either<A, D>) -> Either<A, D> {
    switch self {
    case .Left(let x):
      return Either<A, D>.Left(x)
    case .Right(let x):
      return f(x)
    }
  }
  
  func map<C>(f: (B) -> C) -> Either<A, C> {
    return self.flatMap { Either<A, C>.pure(f($0)) }
  }
}
